File: crates/kuben-platform/src/materializer/detach.rs
//! Detach (M4.11): Kuben lets go of an app and leaves it running.
//!
//! A detach is a `target.delete` that orphans instead of removing:
//! 1. The app's ApplicationRuntime (agent delivery) and App are deleted with
//!    orphan propagation. Before the App goes, it is marked, so its
//!    controller leaves it alone. The garbage collector then drops them from
//!    their workloads' owners, so nothing Kuben runs prunes those workloads
//!    again.
//! 2. The Secrets of the app's last delivered run lose Kuben's
//!    `managed-by` label, so secret garbage collection never takes them.
//! 3. The target is marked deleted, and the detach is marked complete.
//!
//! Volumes were never owned and stay. Routes keep their labels: while Kuben
//! runs, its Gateway keeps serving the detached hostnames. An unreleased
//! detached app also keeps its namespace when the environment is deleted.

use std::time::Duration;

use k8s_openapi::api::core::v1::Secret;
use kube::{
    Api, Resource, ResourceExt,
    api::{DeleteParams, Patch, PatchParams, PropagationPolicy},
};
use kuben_core::ids::{OrgId, TargetId};
use kuben_crd::{App, ApplicationRuntime, labels};
use kuben_store::repo::{Delivery, Subject};
use serde_json::json;

use super::{
    render::annotations,
    worker::{Stop, Worker, refused},
    write,
};

/// How long a detach waits for the garbage collector before looking again.
pub(super) const ORPHAN_WAIT: Duration = Duration::from_secs(2);
/// The label a detached Secret gets instead of Kuben's `managed-by`.
pub const DETACHED: &str = "kuben.dev/detached";

fn orphan() -> DeleteParams {
    DeleteParams {
        propagation_policy: Some(PropagationPolicy::Orphan),
        ..DeleteParams::default()
    }
}

impl Worker {
    /// Take `slug`'s App in `namespace` from its controller and delete it,
    /// orphaning its workloads; done once it is gone. An App that is not
    /// `target`'s (of `org`) is refused.
    pub(super) async fn orphan_app_object(
        &self,
        namespace: &str,
        slug: &str,
        org: OrgId,
        target: TargetId,
    ) -> Result<(), Stop> {
        let apps = Api::<App>::namespaced(self.client.clone(), namespace);
        let Some(live) = apps.get_opt(slug).await? else {
            return Ok(());
        };
        let ours = live
            .annotations()
            .get(annotations::ID)
            .is_none_or(|id| *id == target.to_string());
        if !write::belongs_to(live.meta(), org) || !ours {
            return Err(refused("NameTaken"));
        }
        if !live.annotations().contains_key(annotations::HANDOVER) {
            // First take the App from its controller, which leaves a marked
            // App alone; delete it only after a pause, once a write of that
            // controller already under way has landed. A write after the
            // orphaning would give the workloads back an owner that is going,
            // and the garbage collector would take them with it (CI run
            // 34985518950).
            let mark = json!({ "metadata": { "annotations": {
                annotations::HANDOVER: target.to_string(),
            } } });
            apps.patch(slug, &PatchParams::default(), &Patch::Merge(&mark))
                .await?;
            return Err(Stop::Wait(ORPHAN_WAIT, "HandingOverApp"));
        }
        if live.metadata.deletion_timestamp.is_none() {
            match apps.delete(slug, &orphan()).await {
                Ok(_) => tracing::info!(%target, "the App goes, its workloads stay"),
                Err(kube::Error::Api(s)) if s.code == 404 => return Ok(()),
                Err(e) => return Err(e.into()),
            }
        }
        // The garbage collector drops the App from its workloads' owners,
        // then removes it.
        Err(Stop::Wait(ORPHAN_WAIT, "OrphaningApp"))
    }

    /// Delete `slug`'s ApplicationRuntime, orphaning its workloads; done
    /// once it is gone.
    async fn orphan_runtime(
        &self,
        namespace: &str,
        slug: &str,
        target: TargetId,
    ) -> Result<(), Stop> {
        let runtimes = Api::<ApplicationRuntime>::namespaced(self.client.clone(), namespace);
        let Some(live) = runtimes.get_opt(slug).await? else {
            return Ok(());
        };
        let managed = live
            .labels()
            .get(labels::MANAGED_BY)
            .is_some_and(|v| v == labels::MANAGER);
        let ours = live.spec.target_id == target.to_string();
        if !managed || !ours {
            return Err(refused("NameTaken"));
        }
        if live.metadata.deletion_timestamp.is_none() {
            match runtimes.delete(slug, &orphan()).await {
                Ok(_) => tracing::info!(runtime = slug, "the runtime goes, its workloads stay"),
                Err(kube::Error::Api(s)) if s.code == 404 => return Ok(()),
                Err(e) => return Err(e.into()),
            }
        }
        Err(Stop::Wait(ORPHAN_WAIT, "OrphaningRuntime"))
    }

    /// Carry out the detach of `subject`'s target.
    pub(super) async fn detach_target(&self, org: OrgId, subject: Subject) -> Result<(), Stop> {
        let Some(target) = subject.target else {
            return Err(refused("BadPayload"));
        };
        let mut tenant = self.store.tenant(org).await?;
        let Some(record) = tenant.detached_app(target).await? else {
            return Err(refused("NotDetached"));
        };
        if record.completed_at.is_some() {
            return Ok(()); // finished before
        }
        let delivery = tenant.target_delivery(target).await?;
        let secrets = tenant
            .export_material(target)
            .await?
            .map(|m| m.secrets)
            .unwrap_or_default();
        drop(tenant);
        let (namespace, slug) = (record.namespace.as_str(), record.app.as_str());
        if delivery == Some(Delivery::Agent) {
            self.orphan_runtime(namespace, slug, target).await?;
        }
        self.orphan_app_object(namespace, slug, org, target).await?;
        let api = Api::<Secret>::namespaced(self.client.clone(), namespace);
        let release = json!({ "metadata": { "labels": {
            labels::MANAGED_BY: null,
            DETACHED: target.to_string(),
        } } });
        for secret in &secrets {
            let name = secret.object();
            match api.get_opt(&name).await? {
                Some(live) if write::belongs_to(live.meta(), org) => {
                    api.patch(&name, &PatchParams::default(), &Patch::Merge(&release))
                        .await?;
                }
                _ => {}
            }
        }
        let mut tenant = self.store.tenant(org).await?;
        tenant.finish_target_deletion(subject.project, target).await?;
        tenant.complete_detach(target).await?;
        tenant.commit().await?;
        tracing::info!(%target, app = slug, namespace, secrets = secrets.len(), "app detached");
        Ok(())
    }
}

File: go/hub/internal/platform/materializer/detach.go
package materializer

// Detach (M4.11): Kuben lets go of an app and leaves it running.
//
// A detach is a `target.delete` that orphans instead of removing:
//  1. The app's ApplicationRuntime (agent delivery) and App are deleted
//     with orphan propagation. Before the App goes, it is marked, so its
//     controller leaves it alone. The garbage collector then drops them
//     from their workloads' owners, so nothing Kuben runs prunes those
//     workloads again.
//  2. The Secrets of the app's last delivered run lose Kuben's
//     `managed-by` label, so secret garbage collection never takes them.
//  3. The target is marked deleted, and the detach is marked complete.
//
// Volumes were never owned and stay. Routes keep their labels: while Kuben
// runs, its Gateway keeps serving the detached hostnames. An unreleased
// detached app also keeps its namespace when the environment is deleted.

import (
	"context"
	"time"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ids"
	"github.com/Teamtem-dev/kuben/go/hub/internal/store"
	"github.com/Teamtem-dev/kuben/go/kubenapi/v1alpha1"
)

// orphanWait is how long a detach waits for the garbage collector before
// looking again.
const orphanWait = 2 * time.Second

// LabelDetached is the label a detached Secret gets instead of Kuben's
// `managed-by`.
const LabelDetached = "kuben.dev/detached"

// orphanAppObject takes slug's App in namespace from its controller and
// deletes it, orphaning its workloads; done once it is gone. An App that
// is not target's (of org) is refused.
func (w *Worker) orphanAppObject(ctx context.Context, namespace, slug string, org ids.OrgID, target ids.TargetID) stop {
	ri := w.resource(appsGVR, namespace)
	live, found, err := get[v1alpha1.App](ctx, ri, slug)
	switch {
	case err != nil:
		return retryKube(err)
	case !found:
		return nil
	}
	id, annotated := live.Annotations[v1alpha1.AnnotationID]
	ours := !annotated || id == target.String()
	if !BelongsTo(live.Labels, org) || !ours {
		return refused("NameTaken")
	}
	if _, marked := live.Annotations[v1alpha1.AnnotationHandover]; !marked {
		// First take the App from its controller, which leaves a marked App
		// alone; delete it only after a pause, once a write of that
		// controller already under way has landed. A write after the
		// orphaning would give the workloads back an owner that is going,
		// and the garbage collector would take them with it (CI run
		// 34985518950).
		mark := map[string]any{"metadata": map[string]any{"annotations": map[string]any{
			v1alpha1.AnnotationHandover: target.String(),
		}}}
		if err := mergePatch(ctx, ri, slug, mark); err != nil {
			return retryKube(err)
		}
		return stopWait{after: orphanWait, code: "HandingOverApp"}
	}
	if live.DeletionTimestamp == nil {
		err := ri.Delete(ctx, slug, metav1.DeleteOptions{PropagationPolicy: ptr(metav1.DeletePropagationOrphan)})
		switch {
		case isNotFound(err):
			return nil
		case err != nil:
			return retryKube(err)
		}
		w.d.Logger.Info("the App goes, its workloads stay", "target", target)
	}
	// The garbage collector drops the App from its workloads' owners, then
	// removes it.
	return stopWait{after: orphanWait, code: "OrphaningApp"}
}

// orphanRuntime deletes slug's ApplicationRuntime, orphaning its
// workloads; done once it is gone.
func (w *Worker) orphanRuntime(ctx context.Context, namespace, slug string, target ids.TargetID) stop {
	ri := w.resource(runtimesGVR, namespace)
	live, found, err := get[v1alpha1.ApplicationRuntime](ctx, ri, slug)
	switch {
	case err != nil:
		return retryKube(err)
	case !found:
		return nil
	}
	managed := live.Labels[v1alpha1.LabelManagedBy] == v1alpha1.LabelManagerValue
	if !managed || live.Spec.TargetID != target.String() {
		return refused("NameTaken")
	}
	if live.DeletionTimestamp == nil {
		err := ri.Delete(ctx, slug, metav1.DeleteOptions{PropagationPolicy: ptr(metav1.DeletePropagationOrphan)})
		switch {
		case isNotFound(err):
			return nil
		case err != nil:
			return retryKube(err)
		}
		w.d.Logger.Info("the runtime goes, its workloads stay", "runtime", slug)
	}
	return stopWait{after: orphanWait, code: "OrphaningRuntime"}
}

// detachTarget carries out the detach of subject's target.
func (w *Worker) detachTarget(ctx context.Context, org ids.OrgID, subject store.Subject) stop {
	target, ok := subject.Target.Get()
	if !ok {
		return refused("BadPayload")
	}
	type found struct {
		record   store.DetachedApp
		detached bool
		delivery store.Delivery
		secrets  []store.SecretReference
	}
	f, st := readTenant(ctx, w, org, func(t *store.Tenant) (found, error) {
		var out found
		var err error
		if out.record, out.detached, err = t.DetachedApp(ctx, target); err != nil || !out.detached ||
			out.record.CompletedAt.IsSome() {
			return out, err //nolint:wrapcheck // said as a store error
		}
		if out.delivery, _, err = t.TargetDelivery(ctx, target); err != nil {
			return out, err //nolint:wrapcheck // said as a store error
		}
		material, exported, err := t.ExportMaterial(ctx, target)
		if exported {
			out.secrets = material.Secrets
		}
		return out, err //nolint:wrapcheck // said as a store error
	})
	switch {
	case st != nil:
		return st
	case !f.detached:
		return refused("NotDetached")
	case f.record.CompletedAt.IsSome():
		return nil // finished before
	}
	namespace, slug := f.record.Namespace, f.record.App
	if f.delivery == store.DeliveryAgent {
		if st := w.orphanRuntime(ctx, namespace, slug, target); st != nil {
			return st
		}
	}
	if st := w.orphanAppObject(ctx, namespace, slug, org, target); st != nil {
		return st
	}
	if st := w.releaseSecrets(ctx, namespace, org, target, f.secrets); st != nil {
		return st
	}
	if st := w.writeTenant(ctx, org, func(t *store.Tenant) error {
		if _, err := t.FinishTargetDeletion(ctx, subject.Project, target); err != nil {
			return err //nolint:wrapcheck // said as a store error
		}
		_, err := t.CompleteDetach(ctx, target)
		return err //nolint:wrapcheck // said as a store error
	}); st != nil {
		return st
	}
	w.d.Logger.Info("app detached", "target", target, "app", slug, "namespace", namespace, "secrets", len(f.secrets))
	return nil
}

// releaseSecrets takes Kuben's `managed-by` label off the Secrets of the
// detached app's last delivered run, so secret garbage collection never
// takes them.
func (w *Worker) releaseSecrets(ctx context.Context, namespace string, org ids.OrgID, target ids.TargetID, secrets []store.SecretReference) stop {
	api := w.d.Cluster.Typed.CoreV1().Secrets(namespace)
	release := map[string]any{"metadata": map[string]any{"labels": map[string]any{
		v1alpha1.LabelManagedBy: nil,
		LabelDetached:           target.String(),
	}}}
	for _, secret := range secrets {
		name := secret.Object()
		live, err := api.Get(ctx, name, metav1.GetOptions{})
		switch {
		case isNotFound(err):
			continue
		case err != nil:
			return retryKube(err)
		}
		if live == nil || !BelongsTo(live.Labels, org) {
			continue
		}
		if err := patchSecret(ctx, api, name, release); err != nil {
			return retryKube(err)
		}
	}
	return nil
}

func ptr[T any](v T) *T { return &v }
